src/include/common/worker_pool.h

```diff
@@ -0,0 +1,173 @@
+#pragma once
+#include <atomic>
+#include <condition_variable>  // NOLINT
+#include <functional>
+#include <iostream>
+#include <mutex>  // NOLINT
+#include <queue>
+#include <string>
+#include <thread>  // NOLINT
+#include <utility>
+#include <vector>
+#include "common/macros.h"
+
+namespace terrier::common {
+
+using TaskQueue = std::queue<std::function<void()>>;
+
+/**
+ * A worker pool that maintains a group of worker threads and a task queue.
+ *
+ * As soon as there is a task in the task queue, a worker thread will be
+ * assigned to run that task. A task must be a function that takes no argument.
+ * After a worker finishes a task, it will eagerly try to get a new task.
+ *
+ * This pool is restartable, meaning it can be started again after it has been
+ * shutdown. Calls to Startup() and Shutdown() are thread-safe.
+ */
+class WorkerPool {
+ public:
+  /**
+   * Initialize the worker pool. Once the number of worker is set by the constructor
+   * it cannot be changed.
+   *
+   * After initialization, the worker pool automatically starts up.
+   *
+   * @param num_workers the number of workers in this pool
+   * @param task_queue a queue of tasks
+   */
+  // NOLINTNEXTLINE  lint thinks it has only one arguement
+  WorkerPool(uint32_t num_workers = 1, TaskQueue task_queue = TaskQueue())
```

Comment:
We're avoiding default arguments in the new codebase. TERRIER_ASSERT that it's > 0 in the body of the constructor.